### PR TITLE
Tango support

### PR DESCRIPTION
**Pull request by burkeds**
Enable support for asynchronous control of Tango servers. 

The Tango support backend complies with the Signal object to interface with Tango servers via PyTango device proxies. 

The TangoDevice is an optional device class which inherits from Device. It includes attributes and methods to accommodate PyTango device proxies and to manage device polling for servers that do not support events. TangoReadable is an extension of the StandardReadable class to support Tango.

### CODE REVIEW

**Comment by coretl** [comment]
Before I review this in detail, I thought I would try to describe the type hint approach we use for PandA, and how it could be applied to Tango:

```python
from ophyd_async.core import (
    StandardReadable,
    SignalRW,
    SignalX,
    HintedSignal,
    ConfigSignal,
    DEFAULT_TIMEOUT,
    Device,
)

# This import should go in core after the refactor...
from ophyd_async.epics.pvi import create_children_from_annotations
import inflection


# This goes somewhere in ophyd_async.tango
async def fill_tango_signals(
    device: Device, trl: str, timeout=DEFAULT_TIMEOUT, mock=False
):
    if mock:
        # There is some code in ophyd_async.epics.pvi that we can use for this
        ...
    else:
        # Get a tango proxy and some details about the attributes
        device_proxy = await get_tango_trl(trl)
        configs = await get_tango_configs(trl)
        for attr_trl in device_proxy.get_attribute_list():
            # Convert MyAttribute -> my_attribute
            attr_name = inflection.underscore(attr_trl)
            if not hasattr(device, attr_name):
                # Make a blank signal of the right type, not sure how to tell the RW mode of attributes?
                setattr(device, attr_name, SignalRW())
            # Get the signal off the device and set its backend
            signal = getattr(device, attr_name)
            signal.set_backend(TangoTransport(configs[attr_trl].dtype, device_proxy))
        # Now do similar for commands

    # We call set name now the parent field has been set in all of the
    # introspect-initialized devices. This will recursively set the names.
    device.set_name(device.name)


class SIS3820Counter(StandardReadable):
    counts: SignalRW[float]
    offset: SignalRW[float]
    reset: SignalX

    def __init__(self, trl: str, name: str = "") -> None:
        self._trl = trl
        # Create Signal* objects from annotations but don't set their backends yet
        create_children_from_annotations(self)
        # Register particular signals to contribute to read and read_configuration
        self.add_readables([self.counts], HintedSignal)
        self.add_readables([self.offset], ConfigSignal)
        super.__init__(name=name)

    async def connect(
        self, mock: bool = False, timeout: float = DEFAULT_TIMEOUT
    ) -> None:
        # Fill in the tango backend for the signals we made above, and
        # make new signals for every attribute the tango device made that we didn't mention above
        await fill_tango_signals(self, self._trl, timeout=timeout, mock=mock)
        # Now connect them all in parallel
        await super().connect(mock=mock, timeout=timeout)
```

**Comment by burkeds** [comment]
As of now, tango_signals can automatically infer their types if not specified. I am working on changing TangoReadableDevice::register_signals to automatically instantiate the signals.

**Comment by burkeds** [comment]
@coretl Is there a non-protected method of Signal to set the backend? Currently, to set the backend of the signal I simply do `signal._backend = backend` but it conflicts with Ruff SLF001 to access a protected member of Signal.

**Comment by burkeds** [comment]
@coretl tango devices can now use pytango to automatically infer datatype, signal type, and set a backend. Signals can have the usual `SignalR`,` SignalX`, etc. type hints but `TangoReadableDevice` also supports the typehints `Signal` and `None`. If signals are hinted with `Signal` or `None` then the R/W/X character will also be inferred from the device server and used to set the signal frontend along with the backend.

```python
class TangoMover(TangoReadableDevice, Movable, Stoppable):
    # Enter the name and type of the signals you want to use
    # If type is None or Signal, the type will be inferred from the Tango device
    position: SignalRW
    velocity: SignalRW
    state: SignalR
    _stop: SignalX

    def __init__(self, trl: str, name=""):
        super().__init__(trl, name=name)
        self.add_readables([self.position], HintedSignal)
        self.add_readables([self.velocity], ConfigSignal)
```

**Comment by coretl** [comment]
> @coretl Is there a non-protected method of Signal to set the backend? Currently, to set the backend of the signal I simply do `signal._backend = backend` but it conflicts with Ruff SLF001 to access a protected member of Signal.

No, but you can omit it at signal creation, and supply it at connect instead:
https://github.com/bluesky/ophyd-async/blob/d7f07488fdfa976037791cbdfdb21e0f7c5f7430/src/ophyd_async/core/_signal.py#L61-L67

File: src/ophyd_async/tango/_backend/_tango_transport.py
Comment (coretl) [inline src/ophyd_async/tango/_backend/_tango_transport.py:629]: I don't know much about Tango, are the read_trl and write_trl ever different?

File: src/ophyd_async/tango/_backend/_tango_transport.py
Comment (coretl) [inline src/ophyd_async/tango/_backend/_tango_transport.py:619]: Please can we call this:
```suggestion
class TangoSignalBackend(SignalBackend[T]):
```

File: src/ophyd_async/tango/_backend/_tango_transport.py
Comment (coretl) [inline src/ophyd_async/tango/_backend/_tango_transport.py:1]: Nit: We restructured `ophyd_async.epics.signal` to contain both the `epics_signal_*` helpers, and the CA and PVA `SignalBackend`s. For consistency it would be nice to do the same here and put this in `ophyd_async.tango.signal`

File: src/ophyd_async/tango/demo/counter.py
Comment (coretl) [inline src/ophyd_async/tango/demo/counter.py:42]:
```suggestion
    @AsyncStatus.wrap
    async def trigger(self) -> None:
        sample_time = await self.sample_time.get_value()
        timeout = sample_time + DEFAULT_TIMEOUT
        await self.start.trigger(wait=True, timeout=timeout)
```

File: src/ophyd_async/tango/demo/counter.py
Comment (coretl) [inline src/ophyd_async/tango/demo/counter.py:26]: When we do this in PandA, we specify not just the names of the signals, but their types, e.g.:
```suggestion
    sample_time: SignalRW[float]
```
so that when we `await self.sample_time.get_value()` we know we are getting a float. We then check at connection time that the EPICS PV is actually a float, and error if not. This means we can catch things like `32 + await self.my_string_attribute.get_value()` at lint time rather than runtime

Is it possible to do the same here?

File: src/ophyd_async/tango/base_devices/base_device.py
Comment (coretl) [inline src/ophyd_async/tango/base_devices/base_device.py:26]: I'm wondering if it makes sense to make a `class TangoDevice(Device)`, then have `class TangoMover(TangoDevice, StandardReadable)` which would allow you to make classes that are not `StandardReadable`s (maybe like LIMA plugins?) where they are intended to be children of `StandardDetector` rather than being read in a standalone way?

File: src/ophyd_async/tango/base_devices/base_device.py
Comment (coretl) [inline src/ophyd_async/tango/base_devices/base_device.py:113]: Could you use https://github.com/bluesky/ophyd-async/blob/d7f07488fdfa976037791cbdfdb21e0f7c5f7430/src/ophyd_async/epics/pvi/_pvi.py#L309-L341 if we put it in `ophyd_async.core`?

File: src/ophyd_async/tango/_backend/_tango_transport.py
Comment (coretl) [reply src/ophyd_async/tango/_backend/_tango_transport.py:619]: I think `Transport` is a leftover from when this was in `ophyd.v2`...

File: src/ophyd_async/tango/demo/README.md
Comment (coretl) [inline src/ophyd_async/tango/demo/README.md:49]: Could we wrap this up at the `Device` or `Signal` level? ...

File: src/ophyd_async/tango/demo/mover.py
Comment (coretl) [inline src/ophyd_async/tango/demo/mover.py:34]: ... e.g. I could imagine
```suggestion
@tango_poll_attributes(0.1)
class TangoMover(TangoReadableDevice, Movable, Stoppable):
```
which could set `TangoMover._poll_attribute_rate = 0.1`, then `connect` could pass that information down when it creates the `TangoSignalBackend`.

What do you think?

**Review by coretl — COMMENTED** [review]
Some suggestions, let me know what you think.

Some more code could be shared after #472, although that should be done after this is merged, so it might change too

File: src/ophyd_async/tango/_backend/_tango_transport.py
Comment (burkeds) [reply src/ophyd_async/tango/_backend/_tango_transport.py:629]: I do not know of any examples of this, but yes it is possible. For example, somebody might write a Tango server where `position` is read-only and one must write to `setpoint` or similar to change the position.

**Comment by burkeds** [comment]
@coretl Something that Tango supports is associating Tango commands with typed values. For example, a command like "RunScript" might be a `str` typed command that carries a filepath to a script and returns a string like `Script executed`. Since SignalX only expects `None` typed commands, as a workaround one must treat these commands as SignalRW in ophyd-async. This seems to work ok but it means that ophyd-async will always treat them like attributes. Is there a possibility to introduce typed commands in a future PR?

File: src/ophyd_async/tango/base_devices/base_device.py
Comment (burkeds) [reply src/ophyd_async/tango/base_devices/base_device.py:113]: @coretl this doesn't seem to work at the moment for StandardReadables. This throws an error for _readables because it is a Tuple.

In principle yes, we could use this but it does not allow for inferring the signal frontend and datatype from tango. This is fine but it means we wouldn't be taking advantage of this Tango feature. My instinct would be to overwrite this with a tango version. Perhaps it could accept a callback method to run if you want the frontend to be inferred.

```python
  File "/home/burkedev/projects/BLUESKY/ophyd-async/src/ophyd_async/tango/base_devices/base_device.py", line 50, in __init__
    create_children_from_annotations(self)
  File "/home/burkedev/projects/BLUESKY/ophyd-async/src/ophyd_async/epics/pvi/_pvi.py", line 342, in create_children_from_annotations
    sub_device = device_type()
                 ^^^^^^^^^^^^^
  File "/usr/lib/python3.11/typing.py", line 1246, in __call__
    raise TypeError(f"Type {self._name} cannot be instantiated; "
TypeError: Type Tuple cannot be instantiated; use tuple() instead
```

File: src/ophyd_async/tango/base_devices/base_device.py
Comment (burkeds) [reply src/ophyd_async/tango/base_devices/base_device.py:26]: This makes sense. As suggested I created `TangoDevice(Device)` which contains most of the cost that was previously in `TangoReadableDevice`. `TangoReadableDevice` was redefined to simply inherit from TangoDevice and StandardReadable so that it calls the TangoDevice init method.

File: src/ophyd_async/tango/demo/mover.py
Comment (burkeds) [reply src/ophyd_async/tango/demo/mover.py]: That seems like a good idea. I created a tango_polling class decorator that will disable events and enable polling in this way.

File: src/ophyd_async/tango/demo/README.md
Comment (burkeds) [reply src/ophyd_async/tango/demo/README.md]: Per the other thread, this can now be defined using a class decorator.

File: src/ophyd_async/tango/_backend/_tango_transport.py
Comment (burkeds) [reply src/ophyd_async/tango/_backend/_tango_transport.py:1]: I moved these methods to tango.signal and renamed many of the modules to better resemble the rest of the codebase.

**Comment by burkeds** [comment]
> > @coretl Is there a non-protected method of Signal to set the backend? Currently, to set the backend of the signal I simply do `signal._backend = backend` but it conflicts with Ruff SLF001 to access a protected member of Signal.
> 
> No, but you can omit it at signal creation, and supply it at connect instead:
> 
> https://github.com/bluesky/ophyd-async/blob/d7f07488fdfa976037791cbdfdb21e0f7c5f7430/src/ophyd_async/core/_signal.py#L61-L67

The device level connect calls the connect method of child devices but does not offer a way to pass those children new arguments like backend. It looks like I would have to overwrite Device.connect() entirely.

File: src/ophyd_async/tango/base_devices/_base_device.py
Comment (callumforrester) [inline src/ophyd_async/tango/base_devices/_base_device.py:19]: If possible, I think we should keep the general approach to both epics an tango devices in rough step. That does not mean this PR should squeeze a tango-shaped implementation into an epics-shaped hole. 

For example though, there isn't a generic `EpicsDevice` base class, should there be? Is this a result of a difference between the two control systems or developer-preference? Is there some paradigm we can find where the two implementations look roughly similar, and so the codebase is easier to navigate?

I have picked on this particular example but this comment applies to the whole PR.

File: src/ophyd_async/tango/base_devices/_base_device.py
Comment (callumforrester) [inline src/ophyd_async/tango/base_devices/_base_device.py:91]: Another example of the difference vs. epics (see comment above). Here we make the signal backends in one go in the base class, rather than in the signals themselves. Is this a fundamental requirement of tango or just developer preference? Is there a way we can bring the two implementations more inline?

File: src/ophyd_async/tango/demo/_counter.py
Comment (callumforrester) [inline src/ophyd_async/tango/demo/_counter.py:27]: I think the requirement for type hints is a strength of ophyd-async in general, where we have control systems that are loosely defined, being able to say `counts = Signal[float]` and knowing I will get a sensible error message at connect time if the control system does not actually return a float is very useful. I wouldn't want to see that requirement lost here. See also my other comments about maintaining consistency with the rest of the codebase.

File: src/ophyd_async/tango/demo/README.md
Comment (callumforrester) [inline src/ophyd_async/tango/demo/README.md:1]: This should go in the `docs` directory.

File: src/ophyd_async/tango/base_devices/_base_device.py
Comment (callumforrester) [inline src/ophyd_async/tango/base_devices/_base_device.py:112]: I'm tempted to suggest that polling be per-signal rather than per-device, partly for consistency and partly for flexibility (an ophyd-async device comprising signals from multiple tango servers), but I'm too ignorant about tango servers and common use cases for that to be a firm suggestion.

File: pyproject.toml
Comment (callumforrester) [inline pyproject.toml:57]: Tango should probably have its own extra set like `ca`, `pva` and `sim`

File: src/ophyd_async/tango/signal/_signal.py
Comment (callumforrester) [inline src/ophyd_async/tango/signal/_signal.py:147]: See my other comment about requiring type hints: I would prefer we kept that in ophyd-async. We should require the user to say "The control system should return this type" and not give them the option of saying "This signal is whatever the control system says it is".

File: src/ophyd_async/tango/signal/_signal.py
Comment (callumforrester) [inline src/ophyd_async/tango/signal/_signal.py:189]: This is very nitpicky but I'd prefer these dividers were removed for consistency with the rest of the codebase

**Review by callumforrester — CHANGES_REQUESTED** [review]
The logic here looks good, my concern is keeping the design/structure/API consistent with the rest of `ophyd-async`. I've left a few comments highlighting examples but hopefully the general idea comes across.

File: src/ophyd_async/tango/base_devices/_base_device.py
Comment (coretl) [reply src/ophyd_async/tango/base_devices/_base_device.py:19]: There isn't an `EpicsDevice` because EPICS is decentralized, but there should be a `PviDevice` because there is a single introspection PV which configures the rest of the Device. I think we should concentrate on making `PviDevice` and `TangoDevice` look similar, then we need a `FastCSDevice` that uses the connection mechanism of either of these according to the URI

File: src/ophyd_async/tango/base_devices/base_device.py
Comment (coretl) [reply src/ophyd_async/tango/base_devices/base_device.py]: > @coretl this doesn't seem to work at the moment for StandardReadables. This throws an error for _readables because it is a Tuple.

That should be relatively easy to fix, we should do that

> In principle yes, we could use this but it does not allow for inferring the signal frontend and datatype from tango. This is fine but it means we wouldn't be taking advantage of this Tango feature. My instinct would be to overwrite this with a tango version. Perhaps it could accept a callback method to run if you want the frontend to be inferred.

I'm not sure I understand what the signal frontend is?

As for the datatype, we want to be able to support both specified signals that have specified types, and other unspecified signals that take their type from the control system. For example, imagine a PVI Device that has a `specified_sig` that gives you a `float` when you `caget` it, and an `unspecified_sig` that gives you a `str` when you `caget` it. Then you can:

```python
class MyDevice(PviDevice):
   specified_sig: SignalRW[float]

with DeviceCollector():
    device = MyDevice("PV:PREFIX")
# Now device has both specified_sig and unspecified_sig

await device.specified_sig.put(32.0)  # passes at static analysis and runtime
await device.specified_sig.put("bad")  # fails at static analysis and runtime

RE(save_device(device))  # gets specified_sig and unspecified_sig values
await device.unspecified_sig.put(32.0)  # fails at static analysis but passes at runtime
```

This means that:
- Signals we specify the type of will have their usage checked at static analysis time, then at runtime will be check to ensure the control system type matches
- Signals we don't specify will be added to the device using the control system type. This allows generic load/save to include them, and at runtime you can use the attribute but static analysis will fail if you use it in code
- If you start using an unspecified signal in code, then fix its type and make it a specified signal

File: src/ophyd_async/tango/base_devices/_base_device.py
Comment (burkeds) [reply src/ophyd_async/tango/base_devices/_base_device.py:19]: To expand on what Tom said, there is no EpicsDevice because in EPICS, every PV is basically equivalent to every other PV. There is no inherent signal heirarchy. Tango is fundamentally different from EPICS in this respect because instead of individual signals, you interface with a device server which is intended to be a fully functional device containing attributes, functions, etc. Sort of like an IOC if you were to design one for a single device. This means that a device with a Tango signal requires a shared TangoProxy resource for all signals that are connected to the same TangoServer. Currently, Device does not contain an easy way to share a resource across signals so a new TangoDevice is one way to solve this problem.

File: src/ophyd_async/tango/base_devices/_base_device.py
Comment (burkeds) [reply src/ophyd_async/tango/base_devices/_base_device.py:91]: This choice was made because of the existence of a single tango server for many signals. It makes sense to only create one proxy and pass it to the backend from the Device level than it is to make many equivalent proxies at the signal level.

File: src/ophyd_async/tango/base_devices/_base_device.py
Comment (callumforrester) [reply src/ophyd_async/tango/base_devices/_base_device.py:19]: That makes sense to me, the other option is caching of some sort but I leave it to others to decide what is cleaner

File: src/ophyd_async/tango/base_devices/_base_device.py
Comment (callumforrester) [reply src/ophyd_async/tango/base_devices/_base_device.py:91]: Happy with that, see https://github.com/bluesky/ophyd-async/pull/437#discussion_r1740912296

File: src/ophyd_async/tango/base_devices/_base_device.py
Comment (burkeds) [reply src/ophyd_async/tango/base_devices/_base_device.py:112]: I am not certain on this either. In general, Tango servers can publish events so polling isn't necessary. But some facilities (like DESY) do not use Tango events. Many fast polling signals might effectively DDOS the tango server so one might need to be careful not to do that. The decorator can wrap any class to set the polling attribute so it should work at the device level, or the signal backend. If not, its because I have been inconsistent in how I define the polling attribute. I'll make sure it can work at the device or signal level.

File: src/ophyd_async/tango/signal/_signal.py
Comment (burkeds) [reply src/ophyd_async/tango/signal/_signal.py]: Done.

File: pyproject.toml
Comment (callumforrester) [reply pyproject.toml:57]: Yes, you want it so that `pip install ophyd-async[tango]` installs `pytango` as a dependency, see the `ca` or `pva` section above here, which add dependencies for the case of `pip install ophyd-async[ca]` and `pip install ophyd-async[pva]`

Comment (burkeds) [reply pyproject.toml:57]: Following up on our discussion involving pytango and numpy -- pytango version 10.x will support numpy 2. It should be the next major release.

File: src/ophyd_async/tango/base_devices/_base_device.py
Comment (burkeds) [reply src/ophyd_async/tango/base_devices/_base_device.py:111]: I changed tango_polling to enable either device level with a tuple, signal level polling with a dict, or both.

File: src/ophyd_async/tango/signal/_signal.py
Comment (burkeds) [reply src/ophyd_async/tango/signal/_signal.py]: Annotated signals will now require type hints unless they are SignalX

File: src/ophyd_async/tango/demo/README.md
Comment (burkeds) [reply src/ophyd_async/tango/demo/README.md:1]: Done

File: src/ophyd_async/tango/demo/mover.py
Comment (burkeds) [reply src/ophyd_async/tango/demo/mover.py]: tango_polling can be used as a class decorator to set either device level or signal level polling.

File: src/ophyd_async/tango/base_devices/base_device.py
Comment (burkeds) [reply src/ophyd_async/tango/base_devices/base_device.py]: I write tango-specific solutions to create children from annotations and to fill signals from the tango device server. This will be adapted to a system agnostic solution once one has been finalized.

File: src/ophyd_async/tango/demo/counter.py
Comment (burkeds) [reply src/ophyd_async/tango/demo/counter.py]: Done

File: src/ophyd_async/tango/demo/_detector.py
Comment (coretl) [inline src/ophyd_async/tango/demo/_detector.py:21]: What are these kwargs doing?

I'm adding typing at the moment, and args and kwargs are making things difficult, so if we could remove them that would help...

File: src/ophyd_async/tango/demo/_mover.py
Comment (coretl) [inline src/ophyd_async/tango/demo/_mover.py:79]: Can this be written with `wait_for_value`?
```suggestion
        await wait_for_value(self.state, DevState.ON, timeout=something)
```

**Review by coretl — COMMENTED** [review]
I think this is pretty much there, just a couple of minor points then we can merge. There will probably be some typing things to fix when I've finished turning the type checker back on...

File: src/ophyd_async/tango/demo/_detector.py
Comment (burkeds) [reply src/ophyd_async/tango/demo/_detector.py:21]: Those were there because tango devices need a trl to connect. I refactored the code so that devices can be initialized without a trl but one will have to be set before connecting.

File: src/ophyd_async/tango/demo/_mover.py
Comment (burkeds) [reply src/ophyd_async/tango/demo/_mover.py:79]: Done, that's a handy function.

**Review by coretl — APPROVED** [review]

**Comment by burkeds** [comment]
@callumforrester Are there any other changes you would like to see before merging?

**Review by callumforrester — APPROVED** [review]
LGTM, thanks!

**Comment by coretl** [comment]
@burkeds please could you merge main into this branch once more? `pyright` doesn't seem to be running on the last CI run for some reason...

File: src/ophyd_async/tango/demo/_detector.py
Comment (coretl) [inline src/ophyd_async/tango/demo/_detector.py:13]: If there is no `trl` passed in `__init__`, does it make sense for this to be a `TangoReadable`? It could be a `StandardReadable` instead and not take `trl` in init

File: src/ophyd_async/tango/signal/_signal.py
Comment (coretl) [inline src/ophyd_async/tango/signal/_signal.py:158]: I'd prefer this to be private, so that people didn't end up using `tango_signal_auto` everywhere instead of `tango_signal_r(datatype)`. What do you reckon?

File: src/ophyd_async/tango/base_devices/_tango_readable.py
Comment (burkeds) [reply src/ophyd_async/tango/base_devices/_tango_readable.py:30]: While I don't have a specific use case, it is conceivable that you might have different parts of a tango server in different ophyd devices. Tango servers are not necessarily always written to control a single device. Like ophyd, they can abstract away pretty much anything. For example, you might pass control signals to one ophyd device but scripts or monitoring to another. Or you might have a soft controller device that can send signals to many other devices. This allows you to create one shared resource that can exist across many devices.

File: src/ophyd_async/tango/demo/_detector.py
Comment (burkeds) [reply src/ophyd_async/tango/demo/_detector.py:25]: I think I might have misunderstood the requirements. A single tango device does not carry information about associated devices. For now perhaps we can say that tango will only be able to infer signals.

If your tango database has a device server, its possible that we could infer higher level functionality from that, but a tango server is beyond the scope of this PR.

File: src/ophyd_async/tango/demo/_detector.py
Comment (burkeds) [reply src/ophyd_async/tango/demo/_detector.py:13]: I had kept it as a TangoReadable so it could use the device inference methods in TangoDevice. Since we won't be doing that, I will change it to a StandardReadable

File: src/ophyd_async/tango/signal/_signal.py
Comment (burkeds) [reply src/ophyd_async/tango/signal/_signal.py:158]: We could do that although it may not have the desired effect. By default, device init will infer the existence, types, and RW character of all attributes/commands on a tango device. By not annotating any signals they can simply point a tango device to a TRL and it will automatically do tango_signal_auto for everything. I suspect this will be the case for many users that don't really care about type safety. If they don't care about type safety, they probably also don't care about encapsulation and will just use the method if it is private anyway.

File: src/ophyd_async/tango/demo/_detector.py
Comment (burkeds) [reply src/ophyd_async/tango/demo/_detector.py:25]: It is possible to infer the existence of devices from a tango Database server which IS accessible via a device server API but this is not something I had considered so far. Can implementing this be its own PR in the future?

**Review by coretl — APPROVED** [review]

**Comment by coretl** [comment]
@burkeds please can you merge with a suitable commit message?

**Comment by burkeds** [comment]
> @burkeds please can you merge with a suitable commit message?

@coretl Done. PyTango 10 was released so I updated pyproject to install pytango 10.0.0 or higher. I also had to comment out the unused Device and DeviceVector inferrence methods to improve converage. I left the comments in as a placeholder for this functionality which can hopefully be added in the future.

**Pull request merged**
